docs/proposals/configuration_changes.md

```diff
@@ -0,0 +1,71 @@
+## Configuration Changes
+
+### Goal
+Create a package for creating and managing configuration values as a value store. This will be used to remove the necessity of each package to be loaded just for configuration values. The other benefit is that this package could be a more robust solution to the configuration, Allowing us to do a better job of watching for updates to the configmap and managing subscriptions to those changes. This value store should also be usable by other parts of the application and therefore will require less usage of hard coded configuration structures and will allow the addition or removal of configuration values to not create such an issue. 
+
+### Work Items
+* Create a specific package for managing configuration 
+* Create a package mutex map to be used to store configuration values.
+* Take structured yaml and create a map of map of map.... saving the key as the name of the config value
+```go
+m["broker"] = map["dev_broker": true...]
+m["registry"] = [map["name": "dh"...], map["name": "play"...]]
+```
+* Create a syntax, and document, for retrieving a single configuration value from the list. I.E `config.MustGetBoolConfig("broker.dev_broker")`
+* Create a public API that will give the caller multiple ways to retrieve data.
+    - `MustGet` -> returns a single value and will panic if not able to retrieve the data
+    - `MustGet/Get<Type>` -> Allows for typed returns so the caller does not have to worry about type casting.
+    - Ability to specify a `map[string]interface{}` as a return value so one could say `config.MustGetMapConfig("broker")` and would retrieve the map of values for the broker configuration section. 
+* Update the all of the initialization code to take a map[string]interface{} to create the structures with the correct values. At this point, we can use the "kind" argument for the adapters in the registry addressing [this issue](https://github.com/openshift/ansible-service-broker/issues/49).
+* Allows for the initialization code to print warnings if configuration values are not present, taking care of [this issue](https://github.com/openshift/ansible-service-broker/issues/270).
+* This will also allow us to use the per package logger that we have decided on.
+
+Go Map of generated config
+```go
+map[
+    registry:[
+        map[
+            type:dockerhub 
+            name:dh 
+            url:https://registry.hub.docker.com 
+            user:shurley 
+            pass:testingboom 
+            org:shurley
+        ], 
+        map[
+            pass:testingboom 
+            org:ansibleplaybookbundle 
+            type:dockerhub 
+            name:play 
+            url:https://registry.hub.docker.com 
+            user:shurley
+        ]
+    ] 
+    dao:map[
```

Comment:
This structure seems way more complicated since it's like the specific structs except as maps. Maybe I'm being myopic or simple minded but I always envisioned just a simple flat map.

```yaml
---
registry.docker.type: dockerhub
registry.docker.name: docker
registry.docker.url: https://registry.hub.docker.com
registry.docker.user: DOCKERHUB_USER
registry.docker.pass: DOCKERHUB_PASS
registry.docker.org: DOCKERHUB_ORG
registry.docker.fail_on_error: false
registry.rhcc.type: rhcc
registry.rhcc.name: rhcc
registry.rhcc.url: registry.access.stage.redhat.com
registry.rhcc.fail_on_error: true
registry.rhcc.white_list.1: "^legitimate.*-apb$"
registry.rhcc.black_list.2: "malicious.*-apb$"
registry.rhcc.black_list.3: "^specific-blacklist-apb$"
dao.etcd_host: localhost
dao.etcd_port: 2379
log.logfile: /var/log/ansible-service-broker/asb.log
log.stdout: true
log.level: debug
log.color: true
openshift.host: ""
openshift.ca_file: ""
openshift.bearer_token_file: ""
openshift.apb_sandbox_role: "admin"
openshift.image_pull_policy: IfNotPresent
broker.bootstrap_on_startup: true
broker.dev_broker: true
broker.launch_apb_on_bind: false
broker.recovery: true
broker.output_request: true
broker.ssl_cert_key: /path/to/key
broker.ssl_cert: /path/to/cert
broker.refresh_interval: "600s"
broker.auth.basic.type: basic
broker.auth.basic.enabled: true
```

Then the `Config` struct would look like this:

```golang
type Config interface {
    ToMap() map[string]string
    ContainsKey(key string) bool
    GetBoolean(string, bool) bool
    GetString(key string) string
    GetList(key string) []string
    SubsetMap(prefix string) map[string]string
     ... // many more utility methods
}

type MapConfiguration struct {
    configMap map[string]string
}

func NewConfigurationFromMap(configmap map[string]string) Config {
    return MapConfiguration{configMap: configmap}
}

func NewConfiguration() Config {
    return MapConfiguration{}
}

func (c MapConfiguration) SubsetMap(prefix string) map[string]string {
   // return a new map with all the entries that match the prefix from the embedded configMap
}
```

If I call `config.SubsetMap("registry.docker")` I will get a map with only the following keys:

```
registry.docker.type: dockerhub
registry.docker.name: docker
registry.docker.url: https://registry.hub.docker.com
registry.docker.user: DOCKERHUB_USER
registry.docker.pass: DOCKERHUB_PASS
registry.docker.org: DOCKERHUB_ORG
registry.docker.fail_on_error: false
```

If you called `config.GetBoolean("registry.docker.fail_on_error")` it will return false. Store everything as simple strings.

Reply:
:+1: to the interface 
I think that this would totally work, it is just that we would
1. have to transform the yaml map to this structure
2. For calls like this, we would need to loop through every key to build the new map and return that, while the map approach would be constant time to retrieve each part.
3. returning map[string]string will make the caller use the [strconv](https://golang.org/pkg/strconv/) package, where if it is of type interface, they can just cast the value. I am not opposed to it, just making sure we are all in agreement. 

I am not opposed to doing the extra work in the config package, I just don't know if it will be easier to work with.

Reply:
So after thinking about this overnight and looking at the yaml I proposed, I'm not sure I like it. The original one we had seems better. So -1 to my proposed flat file approach :)

As for map[string]string vs map[string]interface{}. I always struggle with configuration since they're usually just values in a file so treating them all as strings seems logical to me. I guess I fear if we use interface{} you could put in some custom struct in there is that really what we want?

Regarding the caller having to use strconv, I don't understand why? If we offer them conversion methods like `GetInt(key string) int`. For example, if I want the port number for etcd as an int I'd call `config.GetInt("dao.etcd_port")` and get back `2379`. The implementation of GetInt would have to do the conversions.  But if we used `interface{}` no one has to convert. hrm.